File: examples/cmake-advanced/.harmont/pipeline.ts
import { pipeline, push, pullRequest, type PipelineDefinition } from "harmont";
import { cmake } from "harmont/toolchains";

const project = cmake({
  path: ".",
  compiler: "clang-18",
  buildType: "Release",
  std: 20,
  defines: { BUILD_TESTING: "ON" },
});

const sanProject = cmake({ path: ".", compiler: "clang-18" });
const covProject = cmake({ path: "." });

const pipelines: PipelineDefinition[] = [
  {
    slug: "ci",
    triggers: [push({ branch: "main" }), pullRequest()],
    pipeline: pipeline(project.test(), project.lint(), project.fmt(), {
      env: { CI: "true" },
      defaultImage: "ubuntu:24.04",
    }),
  },
  {
    slug: "sanitizers",
    triggers: [push({ branch: "main" })],
    pipeline: pipeline(sanProject.sanitize("asan"), sanProject.sanitize("tsan"), {
      env: { CI: "true" },
      defaultImage: "ubuntu:24.04",
    }),
  },
  {
    slug: "coverage",
    triggers: [push({ branch: "main" })],
    pipeline: pipeline(covProject.coverage(), {
      env: { CI: "true" },
      defaultImage: "ubuntu:24.04",
    }),
  },
];

export default pipelines;

File: examples/cmake-advanced/.harmont/pipeline.py
"""Advanced CMake pipeline — compiler matrix, sanitizers, coverage."""
from __future__ import annotations

import harmont as hm


@hm.pipeline(
    "ci",
    env={"CI": "true"},
    default_image="ubuntu:24.04",
    triggers=[hm.push(branch="main"), hm.pr()],
)
def ci() -> tuple[hm.Step, ...]:
    project = hm.cmake(
        path=".",
        compiler="clang-18",
        build_type="Release",
        std=20,
        defines={"BUILD_TESTING": "ON"},
    )
    return (
        project.test(),
        project.lint(),
        project.fmt(),
    )


@hm.pipeline(
    "sanitizers",
    env={"CI": "true"},
    default_image="ubuntu:24.04",
    triggers=[hm.push(branch="main")],
)
def sanitizers() -> tuple[hm.Step, ...]:
    project = hm.cmake(path=".", compiler="clang-18")
    return (
        project.sanitize("asan"),
        project.sanitize("tsan"),
    )


@hm.pipeline(
    "coverage",
    env={"CI": "true"},
    default_image="ubuntu:24.04",
    triggers=[hm.push(branch="main")],
)
def coverage() -> tuple[hm.Step, ...]:
    project = hm.cmake(path=".")
    return (project.coverage(),)
